src/main/java/jenkins/scm/api/SCMSourceObserver.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2015 CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package jenkins.scm.api;
+
+import hudson.model.Item;
+import hudson.model.TaskListener;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+/**
+ * Callback used by {@link SCMNavigator}.
+ * @since FIXME
+ */
+public abstract class SCMSourceObserver {
+
+    /**
+     * Indicates who is asking for sources.
+     * @return a contextual item, typically a {@code OrganizationFolder}
+     */
+    public abstract @Nonnull SCMSourceOwner getContext();
+
+    /**
+     * Provides a way of reporting progress.
+     * @return a logger
+     */
+    public abstract @Nonnull TaskListener getListener();
+
+    /**
+     * Declare that a new “project” such as a source repository has been found.
+     * @param projectName a name of the project, such as a repository name within an organization; may be used as an {@link Item#getName}
+     * @return a secondary callback to customize the project, on which you must call {@link ProjectObserver#complete}
+     * @throws IllegalArgumentException if this {@code projectName} has already been encountered
+     */
+    public abstract @Nonnull ProjectObserver observe(@Nonnull String projectName) throws IllegalArgumentException;
+
+    /**
+     * Adds extra metadata about the overall organization.
```

Comment:
What happens if you add the same key with different values? is it last wins or will there be multiple values? Can we remove attributes?